File: codeburn/src/providers/device-roots.ts
import { readFileSync, statSync } from 'fs'
import { homedir } from 'os'
import { join, resolve, sep } from 'path'

/// devices.json multi-device roots.
///
/// The aggregator mirrors per-device provider stores under one tree (e.g.
/// ~/codeburn-agg/mirror/<host>/zcode/db.sqlite). devices.json declares which
/// on-disk roots belong to which remote host so session sources can be
/// attributed to a device instead of always reporting the local one:
///
///   [{ "host": "macpro", "zcodeDb": "~/codeburn-agg/mirror/macpro/zcode/db.sqlite",
///      "codexHome": "~/codeburn-agg/mirror/macpro/codex", "workbuddyDb": "..." }]
///
/// zcode/workbuddy sources are cached as `<dbPath>:<sessionId>` (colon
/// suffix), codex sources are file paths under its home directory, so both
/// shapes must prefix-match. Missing/invalid config is not an error: every
/// source simply attributes to the local host.

export type DeviceRoot = {
  host: string
  zcodeDb?: string
  codexHome?: string
  workbuddyDb?: string
}

export function expandHome(p: string): string {
  if (p === '~') return homedir()
  if (p.startsWith('~/')) return join(homedir(), p.slice(2))
  return p
}

function devicesFile(): string {
  return process.env['CODEBURN_DEVICES_FILE'] ?? join(homedir(), '.config', 'codeburn', 'devices.json')
}

// Config rarely changes within a process; re-read only when the file's
// mtime moves. Keyed on (file, mtimeMs) so pointing CODEBURN_DEVICES_FILE at
// a different file always re-reads.
let memo: { file: string; mtimeMs: number; roots: DeviceRoot[] } | null = null

export function loadDeviceRoots(): DeviceRoot[] {
  const file = devicesFile()
  let mtimeMs = 0
  try {
    mtimeMs = statSync(file).mtimeMs
  } catch {
    return []
  }
  if (memo && memo.file === file && memo.mtimeMs === mtimeMs) return memo.roots
  let roots: DeviceRoot[] = []
  try {
    const parsed: unknown = JSON.parse(readFileSync(file, 'utf8'))
    if (Array.isArray(parsed)) {
      roots = parsed
        .filter((d): d is DeviceRoot => !!d && typeof d === 'object' && typeof (d as DeviceRoot).host === 'string')
        .map(d => ({
          host: d.host,
          zcodeDb: d.zcodeDb ? resolve(expandHome(d.zcodeDb)) : undefined,
          codexHome: d.codexHome ? resolve(expandHome(d.codexHome)) : undefined,
          workbuddyDb: d.workbuddyDb ? resolve(expandHome(d.workbuddyDb)) : undefined,
        }))
    }
  } catch {
    roots = []
  }
  memo = { file, mtimeMs, roots }
  return roots
}

export function hostForSourcePath(path: string, localHost: string): string {
  for (const d of loadDeviceRoots()) {
    for (const root of [d.zcodeDb, d.codexHome, d.workbuddyDb]) {
      if (root && (path === root || path.startsWith(root + sep) || path.startsWith(root + ':'))) return d.host
    }
  }
  return localHost
}

File: Sources/WattsonCore/Core/DeviceRoots.swift
// Core/DeviceRoots.swift — devices.json 多设备数据根。
// 聚合器把各设备的 provider 数据镜像在一棵目录树下（如 ~/wattson/mirror/<host>/zcode/db.sqlite）。
// devices.json 声明哪些磁盘路径属于哪台远端主机，使会话数据能归属到设备：
//   [{ "host": "macpro", "zcodeDb": "~/wattson/mirror/macpro/zcode/db.sqlite",
//      "codexHome": "~/wattson/mirror/macpro/codex", "workbuddyDb": "..." }]
// zcode/workbuddy 源缓存键形如 `<dbPath>:<sessionId>`（冒号后缀），codex 源是其
// home 目录下的文件路径，两种形状都要前缀匹配。配置缺失/非法不是错误：一律归属本机。
import Foundation

public struct DeviceRoot: Sendable, Equatable {
    public var host: String
    public var zcodeDb: String?
    public var codexHome: String?
    public var workbuddyDb: String?

    public init(host: String, zcodeDb: String? = nil, codexHome: String? = nil, workbuddyDb: String? = nil) {
        self.host = host
        self.zcodeDb = zcodeDb
        self.codexHome = codexHome
        self.workbuddyDb = workbuddyDb
    }
}

public func devicesFile(env: [String: String] = [:]) -> String {
    if let f = env["WATTSON_DEVICES_FILE"], !f.isEmpty { return f }
    return ((homePath() as NSString).appendingPathComponent(".config/wattson")) + "/devices.json"
}

/// 读取 devices.json（文件缺省/损坏 → 空数组）。进程内很少变化，直接每次读取。
public func loadDeviceRoots(env: [String: String] = [:]) -> [DeviceRoot] {
    let file = devicesFile(env: env)
    guard let parsed = parseJSONFile(file), let list = parsed.arr else { return [] }
    return list.compactMap { item in
        guard let obj = item.obj, let host = obj["host"]?.str else { return nil }
        let resolve = { (p: JSON?) -> String? in
            guard let raw = p?.str, !raw.isEmpty else { return nil }
            return (expandHome(raw) as NSString).standardizingPath
        }
        return DeviceRoot(host: host, zcodeDb: resolve(obj["zcodeDb"]),
                          codexHome: resolve(obj["codexHome"]),
                          workbuddyDb: resolve(obj["workbuddyDb"]))
    }
}

/// 源路径 → 设备名：命中任一远端根前缀则归属该主机，否则本机
public func hostForSourcePath(_ path: String, roots: [DeviceRoot], localHost: String) -> String {
    for d in roots {
        for root in [d.zcodeDb, d.codexHome, d.workbuddyDb].compactMap({ $0 }) {
            if path == root || path.hasPrefix(root + "/") || path.hasPrefix(root + ":") {
                return d.host
            }
        }
    }
    return localHost
}
